### ISSUE TITLE
Create monthly rank list based on correct ranking criteria

### ISSUE THREAD

**Issue by sanjaythiru** [sreeramk28/bccleaderboard#2]
# Ranking criteria for monthly leaderboard
Ranks for the monthly leaderboard are decided based on the following three metrics. The metrics are ordered in decreasing order of priority.
Higher the metric value, better the rank.
Any ties are broken using the same criteria.

1. `clubScoreMonthAggregate`
2. `tournamentPointsMonthAggregate`
3. `tiebreakPointsMonthAggregate`

# TO DO
- Give the same (higher) `tournamentRank` to players with the same `tournamentPoints` AND same `tiebreakPoints`
- Implement the above ranking criteria to create the rank list (the monthly leaderboard)

# Glossary
- `tournamentPoints` - Points scored by a player in a tournament
- `tiebreakPoints` - Tiebreak points secured by a player in a tournament (Based on systems like Buccholz)
- `tournamentRank` - Rank achieved by a player in a tournament. This is based on `tournamentPoints` and `tiebreakPoints` metrics in respective order of priority. Higher the metric value, better the rank
- `clubScore` - Score given to the top 10 players in a tournament by the club. `clubScore = 10 - tournamentRank + 1`
- `clubScoreMonthAggregate` -  Sum of `clubScore` of a player over all tournaments conducted in a month
- `tournamentPointsMonthAggregate` - Sum of `tournamentPoints` of a player over all tournaments conducted in a month
- `tiebreakPointsMonthAggregate` - Sum of `tiebreakPoints`  of a player over all tournaments conducted in a month

**Comment by sanjaythiru** [comment]
Why clubScore = 10 - tournamentRank - 1 ?
( We are actually assigning values from 0 - 9 for ranks 10 - 1 respectively)
In the above scoring scheme, we are not rewarding the 10th rank player with any points.
Thereby treating him one amongst the rest of the players who are not in top 10.

**Comment by sanjaythiru** [comment]
How is lichess currently returning the tournamentRank if the tournamentPoints and tiebreakPoints are the same for 2 players ?
Is the returned tournamentRank same for both the players ?
Or is another criteria used by lichess to tiebreak ?

**Comment by sanjaythiru** [comment]
Overall looks great !

**Comment by sreeramk28** [comment]
> Why clubScore = 10 - tournamentRank - 1 ? ( We are actually assigning values from 0 - 9 for ranks 10 - 1 respectively) In the above scoring scheme, we are not rewarding the 10th rank player with any points. Thereby treating him one amongst the rest of the players who are not in top 10.

Yes there is a mistake. `clubScore = 10 - tournamentRank + 1`. Ranks 1 to 10 will get `clubScore` 10 to 1 respectively.

**Comment by sreeramk28** [comment]
> How is lichess currently returning the tournamentRank if the tournamentPoints and tiebreakPoints are the same for 2 players ? Is the returned tournamentRank same for both the players ? Or is another criteria used by lichess to tiebreak ?

Lichess will see performance rating and give better `tournamentRank` to the player with better performance rating in this case. But we want to give same `tournamentRank`.

**Comment by sanjaythiru** [comment]
![BCC_Dashboard_Sequence_Diagram](https://github.com/sreeramk28/bccleaderboard/assets/43669848/640531cc-521b-42c4-9033-d1d4878452be)

Sequence Diagram

**Comment by sanjaythiru** [comment]
I will work on the class diagram and add.
Please share your thoughts on the above sequence diagram. 
EDIT 1 - I think we can add a Converter between Service and ExtSrcClient which converts the payload from Lichess  / Chess com API to our app entity object.

**Comment by sreeramk28** [comment]
Looks great 💯

**Comment by sreeramk28** [comment]
> I will work on the class diagram and add. Please share your thoughts on the above sequence diagram. EDIT 1 - I think we can add a Converter between Service and ExtSrcClient which converts the payload from Lichess / Chess com API to our app entity object.

For converting the response right? Yes we should add

**Comment by sanjaythiru** [comment]
![BCC_Dashboard_Class_Diagram](https://github.com/sreeramk28/bccleaderboard/assets/43669848/bd5f05ae-baaa-49b2-b765-f3a8810c132d)

Class Diagram

**Comment by sanjaythiru** [comment]
Sreeram, please share your thoughts about the class diagram and sequence diagram. ( i have updated sequence diagram to include conversion also )
Definitely we can make both sequence diagram and class diagram better.
We can use some design patterns to achieve better code implementation which will help us maintain the code.
For Service, I have used factory design pattern.
You can see that the object creation are done in the implementation classes. ( bean is created by spring boot. we can autowire )
